File: common/src/main/java/io/github/notenoughupdates/moulconfig/gui/component/ButtonComponent.java
package io.github.notenoughupdates.moulconfig.gui.component;

import io.github.notenoughupdates.moulconfig.common.KeyboardConstants;
import io.github.notenoughupdates.moulconfig.gui.GuiComponent;
import io.github.notenoughupdates.moulconfig.gui.GuiImmediateContext;
import io.github.notenoughupdates.moulconfig.gui.KeyboardEvent;
import io.github.notenoughupdates.moulconfig.gui.MouseEvent;

public class ButtonComponent extends PanelComponent {
    private final Runnable onClick;

    public ButtonComponent(GuiComponent element, int insets, Runnable onClick, BackgroundRenderer panel) {
        super(element, insets, panel);
        this.onClick = onClick;
    }

    public ButtonComponent(GuiComponent element, int insets, Runnable onClick) {
        this(element, insets, onClick, DefaultBackgroundRenderer.DARK_RECT);
    }

    public Runnable getOnClick() {
        return onClick;
    }

    @Override
    public boolean mouseEvent(MouseEvent mouseEvent, GuiImmediateContext context) {
        if (context.isHovered() && mouseEvent instanceof MouseEvent.Click) {
            MouseEvent.Click click = (MouseEvent.Click) mouseEvent;
            if (click.getMouseState() && click.getMouseButton() == 0) {
                onClick.run();
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean keyboardEvent(KeyboardEvent event, GuiImmediateContext context) {
        if (isFocused() && event instanceof KeyboardEvent.KeyPressed) {
            KeyboardEvent.KeyPressed keyPressed = (KeyboardEvent.KeyPressed) event;
            if (keyPressed.getPressed() && keyPressed.getKeycode() == KeyboardConstants.INSTANCE.getEnter()) {
                onClick.run();
                return true;
            }
        }
        return false;
    }
}

File: common/src/main/java/io/github/notenoughupdates/moulconfig/gui/component/ButtonComponent.kt
package io.github.notenoughupdates.moulconfig.gui.component

import io.github.notenoughupdates.moulconfig.common.KeyboardConstants
import io.github.notenoughupdates.moulconfig.gui.GuiComponent
import io.github.notenoughupdates.moulconfig.gui.GuiImmediateContext
import io.github.notenoughupdates.moulconfig.gui.KeyboardEvent
import io.github.notenoughupdates.moulconfig.gui.MouseEvent
import io.github.notenoughupdates.moulconfig.gui.MouseEvent.Click

class ButtonComponent(
    element: GuiComponent,
    insets: Int,
    val onClick: Runnable
) : PanelComponent(element, insets, DefaultBackgroundRenderer.DARK_RECT) {
    override fun mouseEvent(mouseEvent: MouseEvent, context: GuiImmediateContext): Boolean {
        if (context.isHovered && mouseEvent is Click) {
            val (mouseButton, mouseState) = mouseEvent
            if (mouseState && mouseButton == 0) {
                onClick.run()
                return true
            }
        }
        return false
    }

    override fun keyboardEvent(event: KeyboardEvent, context: GuiImmediateContext): Boolean {
        if (isFocused && event is KeyboardEvent.KeyPressed &&
            event.pressed && event.keycode == KeyboardConstants.enter
        ) {
            onClick.run()
            return true
        }
        return false
    }
}
